### DISCUSSION TITLE
Allow assemblies to have types default to sealed

### DISCUSSION THREAD

**Discussion by NorthernLightsDevel** [dotnet/csharplang#6506] [Language Ideas]
Proposing adding a new attribute to the csproj file e.g.
``` xml
<PropertyGroup>
   <Sealed>enable/disable</Sealed>
</PropertyGroup>
```
This change would alter the behavior of the compiler to implicitly mark classes as sealed if not otherwise marked, meaning they will not be allowed to be inherited and cannot have virtual members.
e.g.:
``` CSharp
public class Example ... // Equivalent of public sealed class ... if Sealed=enabled
{}
public class Example2 : Example ... // Not allowed if Sealed=enabled
{}
```
This would require a change to the allowed modifiers of types to allow the `virtual` modifier.

``` CSharp
public virtual class Example ...  // Equivalent of public class Example if Sealed is missing or disabled.
{}

public class Example2 : Example ... // Always allowed since Example is virtual.
{}

```
Virtual members should be allowed in all virtual and abstract types the current behavior, but when Sealed=enabled, types without a modifier will not be allowed virtual members.

**Comment by CyrusNajmabadi** [comment]
this is very unlikely to happen.  NRT was an extremely special case where the benefits were enormous, and the need for runtime compat with teh entire ecosystem was paramount.  Because of htat, we allowed this special exception and added flags.

For something like this, the best path forward would be to write an analyzer that errors on non-sealed classes *unless* it has a special attribute on it you define (e.g. `[AllowUnsealed]`) to indicate you're ok with this particular class being non-sealed.  Such an analyzer would only be a handful of lines of code, could ship today, and could work down-level on many years' worth of compilers.

**Reply by jnm2** [reply]
If it was valid for only one class declaration part to actually have the `partial` keyword, source generators could seal all classes by default.

**Reply by BenMakesGames** [reply]
I agree that `sealed`-by-default isn't _as_ incredible as nullable reference types, but I think it's still a great move.

* `sealed`-by-default doesn't prevent bugs in the same way NRT does, but like NRT, it encourages better design/architecture in common-task code (non-frameworks)
  * personal note: telling juniors to stop trying to solve problems with inheritance all the time is an uninteresting part of work. anything to help here is welcome :P
* `sealed`-by-default sprinkles some minor performance improvements, so that's nice, too :D
* Like NRT, it's too late for the language to be changed; allowing opt-in on a per-project basis is (probably?) the only option, so configuration via csproj feels like a good fit.

**Reply by CyrusNajmabadi** [reply]
> the only option, so configuration via csproj feels like a good fit.

Another feasible option, that can be done right now with almost no work, is to just write an analyzer for this. :-)

**Reply by 333fred** [reply]
ChatGPT can actually produce this analyzer too.

**Reply by CyrusNajmabadi** [reply]
Dey Terk Er Jerbs

**Comment by jnm2** [comment]
Consider the built-in CA1852 diagnostic for this that is shipping in the .NET 7 SDK: https://github.com/dotnet/runtime/issues/49944

**Comment by Mrxx99** [comment]
https://www.nuget.org/packages/Seal.Fody
